### PR TITLE
Add support for Amlogic s905x (GXL)

### PR DESCRIPTION
**Pull request by repk**
This patchset adds support for Amlogic s905x aka GXL boards. It has been tested on a lepotato board and boot up a Linux fine.

### CODE REVIEW

**Comment by antonio-nino-diaz-arm** [comment]
So there are a couple of issues with this. The main one is that even though most files are called GXL, most functions and definitions are called GXBB because... well, when I started this port I didn't bother renaming them. My idea was to try to make common as much code as possible between GXBB and GXL because there are only a few differences, but I gave up when I couldn't get the SCP to work. I think that for this PR we don't really need to modify the ports in order to share code, but it's something to keep in mind (unless you want to do it, feel free to do it in that case).

You're also missing the signed-off lines in your commit messages.

The last thing: try to build with ``ERROR_DEPRECATED=1``, and maybe update the document ``docs/plat/meson-gxbb.rst`` (maybe rename it?) to add instructions for this platform.

**Comment by repk** [comment]
Yes I was planning to mutualize common code between gxbb and gxl. I'll try to do that within next couple of days.

**Comment by repk** [comment]
Also what is the rule here, should sha_dma.c go into plat/meson/gxl/drivers/crypto or drivers/meson/gxl/crypto/ ? I see both are used, is there any preferences on that ?

**Comment by antonio-nino-diaz-arm** [comment]
``drivers/meson/gxl/crypto`` is the preferred one.

File: tools/meson/doimage.c [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+ */
+#include <stdlib.h>
+#include <stdio.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <stdint.h>
+#include <endian.h>
+
+#define DEFAULT_PROGNAME "doimage"
+#define PROGNAME(argc, argv) ((argc >= 1) ? ((argv)[0]) : DEFAULT_PROGNAME)
```
Comment (antonio-nino-diaz-arm) [inline tools/meson/doimage.c]: Add () to argc.

**Comment by repk** [comment]
Oups sorry about your sign off, I have even messed up mine.

File: docs/plat/meson-gxl.rst [hunk 2]
```diff
@@ -0,0 +1,27 @@
+Trusted Firmware-A for Amlogic Meson S905x (GXL)
```
Comment (antonio-nino-diaz-arm) [inline docs/plat/meson-gxl.rst]: Update the list of paths in ``maintainers.rst`` and add this platform to the list of supported platforms in the readme.

File: include/drivers/meson/gxl/crypto/sha_dma.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2019, Remi Pommarel <repk@triplefau.lt>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#ifndef _SHA_DMA_H_
```
Comment (antonio-nino-diaz-arm) [inline include/drivers/meson/gxl/crypto/sha_dma.h]: Just ``SHA_DMA_H``.

File: plat/meson/gxl/gxl_sip_svc.c [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	case GXBB_SM_JTAG_ON:
+		scpi_jtag_set_state(GXBB_JTAG_STATE_ON, x1);
+		SMC_RET1(handle, 0);
+
+	case GXBB_SM_JTAG_OFF:
+		scpi_jtag_set_state(GXBB_JTAG_STATE_OFF, x1);
+		SMC_RET1(handle, 0);
+
+	//case 0x820000FF:
+	//	SMC_RET1(handle, 0);
```
Comment (antonio-nino-diaz-arm) [inline plat/meson/gxl/gxl_sip_svc.c]: Remove this comment.

File: plat/meson/gxl/platform.mk [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+all: image ${BUILD_PLAT}/bl31.img
+distclean realclean clean: cleanimage
+
+cleanimage:
+	${Q}${MAKE} -C ${DOIMAGEPATH} clean
+
+image:
+	${Q}${MAKE} -C ${DOIMAGEPATH}
+
+${BUILD_PLAT}/bl31.img: ${BUILD_PLAT}/bl31.bin ${DOIMAGETOOL}
```
Comment (antonio-nino-diaz-arm) [inline plat/meson/gxl/platform.mk]: You need to add ``image`` to the list of dependencies of ``${BUILD_PLAT}/bl31.img``.

**Comment by antonio-nino-diaz-arm** [comment]
Try to build with ``CROSS_COMPILE=aarch64-linux-gnu- make -j PLAT=gxl ERROR_DEPRECATED=1`` and fix all the includes. If you don't, the platform will break when this patch is applied: https://review.trustedfirmware.org/#/c/TF-A/trusted-firmware-a/+/756/

**Comment by repk** [comment]
I have done that in the meantime, sorry.

**Comment by antonio-nino-diaz-arm** [comment]
Apart from this, everything else looks fine to me.

**Comment by repk** [comment]
Btw, if you prefer to merge only with the common code mutualization between gxbb and gxl, this PR could wait a couple of days.

**Comment by antonio-nino-diaz-arm** [comment]
Personally I don't mind. The code of this platform is relatively small so I don't think it's a big problem to have duplicated code for a few days.

IP Review +1

**Comment by antonio-nino-diaz-arm** [comment]
jenkins: test this please

File: docs/plat/meson-gxl.rst [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+::
+
+    CROSS_COMPILE=aarch64-linux-gnu- make DEBUG=1 PLAT=gxl
+
+This port has been tested on a Lepotato. After building it, follow the
+instructions in the `gxlimg repository` or `U-Boot repository`_, replacing the
+mentioned **bl31.img** by the one built from this port.
+
+.. _gxlimg repository: https://github.com/repk/gxlimg/blob/master/README
+.. _U-Boot repository: https://github.com/u-boot/u-boot/blob/master/board/amlogic/odroid-c2/README.odroid-c2
```
Comment (antonio-nino-diaz-arm) [inline docs/plat/meson-gxl.rst]: This is linking to the Odroid-C2 readme, it should point at https://github.com/u-boot/u-boot/blob/master/board/amlogic/p212/README.libretech-cc, right?

**Comment by antonio-nino-diaz-arm** [comment]
I've just tested it on my lepotato board, and it works. However, there is one more thing to do:
```
[WARNING: BL31: cortex_a53: CPU workaround for 819472 was missing!
WARNING: BL31: cortex_a53: CPU workaround for 824069 was missing!
VERBOSE: BL31: cortex_a53: CPU workaround for 826319 was not applied
WARNING: BL31: cortex_a53: CPU workaround for 827319 was missing!
VERBOSE: BL31: cortex_a53: CPU workaround for 835769 was not applied
VERBOSE: BL31: cortex_a53: CPU workaround for disable_non_temporal_hint was not applied
VERBOSE: BL31: cortex_a53: CPU workaround for 843419 was not applied
INFO:    BL31: cortex_a53: CPU workaround for 855873 was applied
```
You need to enable the errata workarounds that say "was missing". Just go to your ``platform.mk`` and add this:
```
ERRATA_A53_819472		:= 1
ERRATA_A53_824069		:= 1
ERRATA_A53_827319		:= 1
```

File: docs/plat/meson-gxl.rst (hunk 6 above)
Comment (repk) [reply docs/plat/meson-gxl.rst]: Ah good catch

**Comment by repk** [comment]
Halt works better than vendor's one though. I'll take a look at the reboot issue.

**Comment by antonio-nino-diaz-arm** [comment]
I wouldn't bother too much with the reboot issue, as far as I can see SCP sees the command, but it fails to do the reboot. Maybe it's because there are secondary cores on, who knows, but that's not a blocker for this PR, that could be fixed if you actually discover the cause. Considering that CPU0 can't be powered off like the other CPUs, I'd say that a but in the reboot command isn't unlikely.

In any case, I'm happy with the PR.

**Comment by antonio-nino-diaz-arm** [comment]
jenkins: test this please

**Pull request merged**